adapters/webhdfs-adapter/src/main/java/com/connexta/replication/adapters/webhdfs/WebHdfsNodeAdapter.java

```diff
@@ -126,7 +151,205 @@ public String getSystemName() {
+            METADATA_ATTRIBUTE_TYPE_STRING,
+            Collections.singletonList(String.valueOf(fileStatus.getLength())),
+            Collections.emptyList());
+    metadataAttributes.put(Core.RESOURCE_SIZE, lengthAttribute);
+
+    Metadata metadata = new MetadataImpl(metadataAttributes, Map.class, id, modificationTime);
+
+    try {
+      metadata.setResourceUri(new URI(fileUrl));
+      metadata.setResourceModified(modificationTime);
+      metadata.setResourceSize(fileStatus.getLength());
+
+      return metadata;
+    } catch (URISyntaxException e) {
+      throw new ReplicationException("Unable to create a URI from the file's URL.", e);
+    }
+  }
+
+  /**
+   * Using the file URL and modification time as input, a version-3 UUID is generated which is then
+   * modified to appear as a version-4 UUID by changing the version number in the UUID from 3 to 4.
+   * The URL is not guaranteed to be unique if a file is removed and replaced by another file with
+   * the same URL. Because of this, the URL is combined with the modification time to serve as input
+   * for UUID generation.
+   *
+   * @param fileUrl the full URL of the file
+   * @param modificationTime modification time of the file
+   * @return a {@code String} representing a version-4 UUID
+   */
+  private String getVersion4Uuid(String fileUrl, Date modificationTime) {
+    String input = String.format("%s_%d", fileUrl, modificationTime.getTime());
+    String version3Uuid = UUID.nameUUIDFromBytes(input.getBytes()).toString();
+    StringBuilder version4Uuid = new StringBuilder(version3Uuid);
+    version4Uuid.setCharAt(UUID_VERSION_INDEX, '4');
+
+    return version4Uuid.toString();
+  }
+
+  /**
+   * Formulates a GET request to send to the HDFS instance. Repeats the request to retrieve
+   * additional results if the file system contains more results than can be returned in a single
+   * response.
+   *
+   * @param filterDate specifies a point in time such that only files more recent are returned
+   * @return a resulting {@code List} of {@link FileStatus} objects meeting the criteria
+   */
+  @VisibleForTesting
+  List<FileStatus> getFilesToReplicate(@Nullable Date filterDate) {
+
+    List<FileStatus> filesToReplicate = new ArrayList<>();
+    AtomicInteger remainingEntries = new AtomicInteger();
+    AtomicReference<String> pathSuffix = new AtomicReference<>();
+
+    do {
+      try {
+        URIBuilder builder = new URIBuilder(getWebHdfsUrl().toString());
+        builder.setParameter(HTTP_OPERATION_KEY, HTTP_OPERATION_LIST_STATUS_BATCH);
+
+        if (pathSuffix.get() != null) {
+          builder.setParameter(HTTP_START_AFTER_KEY, pathSuffix.get());
+        }
+
+        HttpGet httpGet = new HttpGet(builder.build());
+
+        ResponseHandler<List<FileStatus>> responseHandler =
+            response -> {
+              int statusCode = response.getStatusLine().getStatusCode();
+
+              if (statusCode == HttpStatus.SC_OK) {
```

Comment:
If you're looking to add some logging, it might be useful to have DEBUG messages in the responses so that we can see how what requests are passing/failing. You could add the logging here and in other methods with a response handler

Reply:
That's fair, although these usually throw exceptions with info when it is not the expected status code returned.